acracensor/handlers/querycapture_handler.go

```diff
@@ -5,89 +5,127 @@ import (
 	if err != nil {
 		return nil, err
 	}
 
-	emptyQuery := &QueryInfo{RawQuery:"", IsForbidden:false}
-	occuredError := &CaptureError{err:nil, query:emptyQuery}
+	var queries []QueryInfo
 
-	logChannel_ := make(chan QueryInfo)
+	if len(bufferBytes) != 0 {
+		if err = json.Unmarshal(bufferBytes, &queries); err != nil{
+			return nil, err
+		}
+	}
 
+	logChannel := make(chan QueryInfo, MaxQueriesInChannel)
+	signalToSerialize := make(chan bool)
+	signalShutdown := make(chan os.Signal, 2)
+	signal.Notify(signalShutdown, os.Interrupt, syscall.SIGTERM)
+
+	//serialization signal
+	go func() {
+		for {
+			time.Sleep(TimeoutSecondsToSerialize * time.Second)
+			//timer finished. Close channel to inform that serialization should be performed
+			signalToSerialize <- true
+		}
+	}()
+
+	handler := &QueryCaptureHandler{}
+	handler.filePath = filePath
+	handler.Queries = queries
+	handler.signalToSerialize = signalToSerialize
+	handler.logChannel = logChannel
+
+	//handling goroutine
 	go func (){
 		for {
 			select {
-			case queryInfo, gotQuery := <-logChannel_:
-				if gotQuery {
+			case <-signalToSerialize:
+				err := handler.Serialize()
+				if err != nil {
+					log.WithError(ErrComplexSerializationError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
+				}
+
+			case queryInfo, ok := <-handler.logChannel:
+				if ok {
 					bytes, err := json.Marshal(queryInfo)
 					if err != nil {
-						occuredError.query = &queryInfo
-						occuredError.err = err
+						log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
 					}
 
 					f, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
 					if err != nil {
-						occuredError.query = &queryInfo
-						occuredError.err = err
+						log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
 					}
 
-					defer f.Close()
 					if _, err = f.WriteString("\n"); err != nil {
-						occuredError.query = &queryInfo
-						occuredError.err = err
+						log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
 					}
 
 					if _, err = f.Write(bytes); err != nil {
-						occuredError.query = &queryInfo
-						occuredError.err = err
+						log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
 					}
+					f.Close()
+
 				} else {
-					//channel is closed
-					return
+					//channel is unexpectedly closed
+					log.WithError(ErrUnexpectedCaptureChannelClose).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
+				}
+			case <-signalShutdown:
+				err := handler.Serialize()
+				if err != nil {
+					log.WithError(ErrComplexSerializationError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
 				}
```

Comment:
add return to stop goroutine

Reply:
Do we need to return? As we discussed with @shadinua and @vixentael if error in background occurred, we should inform about this and continue working. Am I missing something?

Reply:
it was releated to standard processing (`case queryInfo, ok := <-handler.logChannel`). That when you took any errors, you should continue with some delay. But you should add some way to stop this goroutine. For example for tests, or restarting (for example on changes from webconfig), or something else.